File: 303 - Range Sum Query - Immutable/cpp/ans.cpp
// Method 1
class NumArray {
public:
    vector<int> nums;
    NumArray(vector<int>& nums) {
        this->nums = nums;
    }
    
    int sumRange(int left, int right) {
        return accumulate(std::next(this->nums.begin(), left), std::next(this->nums.begin(), right + 1), 0);
    }
};

// Method 2
class NumArray {
public:
    vector<int> sums;
    NumArray(vector<int>& nums) {
        int current_sum = 0;
        for(auto num: nums){
            current_sum += num;
            this->sums.push_back(current_sum);
        }
    }
    
    int sumRange(int left, int right) {
        if (left == 0) {
            return this->sums[right];
        } else {
            return this->sums[right] - this->sums[left - 1];
        }
    }
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray* obj = new NumArray(nums);
 * int param_1 = obj->sumRange(left,right);
 */

File: 303 - Range Sum Query - Immutable/go/ans.go
// Method 1
type NumArray struct {
    Nums []int
}


func Constructor(nums []int) NumArray {
    return NumArray{Nums: nums}
}


func (this *NumArray) SumRange(left int, right int) int {
    sums := 0
    for i := left; i <= right; i++ {
        sums += this.Nums[i]
    }
    return sums
}

// Method 2
type NumArray struct {
    Sums []int
}


func Constructor(nums []int) NumArray {
    sums := []int{}
    current_sum := 0
    for _, num := range nums {
        current_sum += num
        sums = append(sums, current_sum)
    }
    return NumArray{Sums: sums}
}


func (this *NumArray) SumRange(left int, right int) int {
    if left == 0 {
        return this.Sums[right]
    } else {
        return this.Sums[right] - this.Sums[left - 1]
    }
}


/**
 * Your NumArray object will be instantiated and called as such:
 * obj := Constructor(nums);
 * param_1 := obj.SumRange(left,right);
 */